File: apps/mobile-app/ios/VaultUtils/TotpGenerator.swift
import Foundation
import CommonCrypto

/// Utility class for generating Time-based One-Time Passwords (TOTP) codes.
/// Implements RFC 6238 TOTP algorithm.
public class TotpGenerator {
    /// The HMAC algorithms supported by the generator, mapped to their CommonCrypto identity.
    /// An unrecognized name falls back to SHA1, which is what RFC 6238 assumes.
    private static func hmacAlgorithm(for name: String) -> (algorithm: CCHmacAlgorithm, digestLength: Int) {
        switch name.uppercased() {
        case "SHA256":
            return (CCHmacAlgorithm(kCCHmacAlgSHA256), Int(CC_SHA256_DIGEST_LENGTH))
        case "SHA512":
            return (CCHmacAlgorithm(kCCHmacAlgSHA512), Int(CC_SHA512_DIGEST_LENGTH))
        default:
            return (CCHmacAlgorithm(kCCHmacAlgSHA1), Int(CC_SHA1_DIGEST_LENGTH))
        }
    }

    /// Generate a TOTP code from a secret key.
    /// - Parameters:
    ///   - secret: The Base32-encoded secret key
    ///   - time: The current time (defaults to now)
    ///   - period: The time step in seconds (defaults to 30)
    ///   - digits: Number of digits in the code (defaults to 6)
    ///   - algorithm: The HMAC algorithm name — SHA1, SHA256 or SHA512 (defaults to SHA1)
    /// - Returns: The generated TOTP code as a string, or nil if generation fails
    public static func generateCode(
        secret: String,
        time: Date = Date(),
        period: Int = 30,
        digits: Int = 6,
        algorithm: String = "SHA1"
    ) -> String? {
        // Decode the Base32 secret
        guard let secretData = base32Decode(secret) else {
            return nil
        }

        // A zero or negative period would divide by zero; treat it as the standard step.
        let step = period > 0 ? period : 30

        // Calculate the time counter
        let counter = UInt64(time.timeIntervalSince1970) / UInt64(step)

        // Generate HOTP code
        return generateHOTP(secret: secretData, counter: counter, digits: digits, algorithm: algorithm)
    }

    /// Generate an HMAC-based One-Time Password (HOTP) code.
    /// - Parameters:
    ///   - secret: The secret key as Data
    ///   - counter: The counter value
    ///   - digits: Number of digits in the code
    ///   - algorithm: The HMAC algorithm name — SHA1, SHA256 or SHA512
    /// - Returns: The generated HOTP code as a string
    private static func generateHOTP(secret: Data, counter: UInt64, digits: Int, algorithm: String) -> String {
        // Convert counter to 8-byte big-endian
        var counterBytes = counter.bigEndian
        let counterData = Data(bytes: &counterBytes, count: 8)

        // Calculate the HMAC using the requested algorithm
        let hmacSpec = hmacAlgorithm(for: algorithm)
        var hmac = [UInt8](repeating: 0, count: hmacSpec.digestLength)
        secret.withUnsafeBytes { secretBytes in
            counterData.withUnsafeBytes { counterBytes in
                CCHmac(
                    hmacSpec.algorithm,
                    secretBytes.baseAddress,
                    secret.count,
                    counterBytes.baseAddress,
                    counterData.count,
                    &hmac
                )
            }
        }

        // Dynamic truncation
        let offset = Int(hmac[hmac.count - 1] & 0x0F)
        let truncatedHash = hmac[offset..<offset + 4]

        // Convert to UInt32 and apply mask
        var code: UInt32 = 0
        for byte in truncatedHash {
            code = (code << 8) | UInt32(byte)
        }
        code &= 0x7FFFFFFF

        // Generate the final code with specified number of digits
        let modulo = UInt32(pow(10.0, Double(digits)))
        let otp = code % modulo

        // Format with leading zeros
        return String(format: "%0\(digits)d", otp)
    }

    /// Decode a Base32-encoded string.
    /// - Parameter string: The Base32-encoded string
    /// - Returns: The decoded Data, or nil if decoding fails
    private static func base32Decode(_ string: String) -> Data? {
        // Remove spaces and convert to uppercase
        let cleanedString = string
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "-", with: "")
            .uppercased()

        // Base32 alphabet (RFC 4648)
        let base32Alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

        var bits = ""

        // Convert each character to its 5-bit binary representation
        for char in cleanedString {
            guard let index = base32Alphabet.firstIndex(of: char) else {
                // Invalid character in Base32 string
                return nil
            }
            let value = base32Alphabet.distance(from: base32Alphabet.startIndex, to: index)
            bits += String(value, radix: 2).padLeft(toLength: 5, withPad: "0")
        }

        // Convert binary string to bytes
        var bytes = [UInt8]()
        for iVal in stride(from: 0, to: bits.count, by: 8) {
            let endIndex = min(iVal + 8, bits.count)
            let byteBits = bits[bits.index(bits.startIndex, offsetBy: iVal)..<bits.index(bits.startIndex, offsetBy: endIndex)]
            if byteBits.count == 8 {
                if let byte = UInt8(byteBits, radix: 2) {
                    bytes.append(byte)
                }
            }
        }

        return Data(bytes)
    }

    /// Get the remaining seconds until the next TOTP code.
    /// - Parameters:
    ///   - time: The current time (defaults to now)
    ///   - period: The time step in seconds (defaults to 30)
    /// - Returns: The number of seconds remaining
    public static func getRemainingSeconds(time: Date = Date(), period: Int = 30) -> Int {
        let step = period > 0 ? period : 30
        let elapsed = Int(time.timeIntervalSince1970) % step
        return step - elapsed
    }
}

/// String extension for padding.
private extension String {
    func padLeft(toLength length: Int, withPad pad: String) -> String {
        let padLength = length - self.count
        if padLength <= 0 {
            return self
        }
        return String(repeating: pad, count: padLength) + self
    }
}

File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/utils/TotpGenerator.kt
package net.aliasvault.app.utils

import android.util.Log
import java.nio.ByteBuffer
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.experimental.and

/**
 * RFC 6238 TOTP generator.
 */
object TotpGenerator {
    private const val TAG = "TotpGenerator"
    private const val DEFAULT_PERIOD = 30
    private const val DEFAULT_DIGITS = 6
    private const val DEFAULT_ALGORITHM = "SHA1"

    /**
     * Map a TOTP algorithm name to its JCA Mac name. An unrecognized name falls back to
     * HmacSHA1, which is what RFC 6238 assumes.
     */
    private fun macName(algorithm: String): String = when (algorithm.uppercase()) {
        "SHA256" -> "HmacSHA256"
        "SHA512" -> "HmacSHA512"
        else -> "HmacSHA1"
    }

    /**
     * Generate the current TOTP code for a Base32-encoded secret.
     * Returns null when the secret is invalid or HMAC fails.
     */
    fun generateCode(
        secret: String,
        timeSeconds: Long = System.currentTimeMillis() / 1000L,
        period: Int = DEFAULT_PERIOD,
        digits: Int = DEFAULT_DIGITS,
        algorithm: String = DEFAULT_ALGORITHM,
    ): String? {
        val secretBytes = base32Decode(secret) ?: return null
        if (secretBytes.isEmpty()) return null

        // A zero or negative period would divide by zero; treat it as the standard step.
        val step = if (period > 0) period else DEFAULT_PERIOD
        val counter = timeSeconds / step
        return generateHotp(secretBytes, counter, digits, algorithm)
    }

    private fun generateHotp(secret: ByteArray, counter: Long, digits: Int, algorithm: String): String? {
        return try {
            val counterBytes = ByteBuffer.allocate(Long.SIZE_BYTES).putLong(counter).array()

            val mac = Mac.getInstance(macName(algorithm))
            mac.init(SecretKeySpec(secret, macName(algorithm)))
            val hash = mac.doFinal(counterBytes)

            // Dynamic truncation per RFC 4226
            val offset = (hash[hash.size - 1] and 0x0F).toInt()
            val binary = ((hash[offset].toInt() and 0x7F) shl 24) or
                ((hash[offset + 1].toInt() and 0xFF) shl 16) or
                ((hash[offset + 2].toInt() and 0xFF) shl 8) or
                (hash[offset + 3].toInt() and 0xFF)

            val mod = pow10(digits)
            val code = binary % mod
            code.toString().padStart(digits, '0')
        } catch (e: Exception) {
            Log.w(TAG, "TOTP HMAC generation failed", e)
            null
        }
    }

    private fun pow10(n: Int): Int {
        var result = 1
        repeat(n) { result *= 10 }
        return result
    }

    /**
     * Decode a Base32 (RFC 4648) string to bytes. Tolerates lowercase, spaces,
     * and trailing padding ('='). Returns null on invalid characters.
     */
    private fun base32Decode(input: String): ByteArray? {
        val cleaned = input.uppercase().replace(" ", "").replace("=", "")
        if (cleaned.isEmpty()) return ByteArray(0)

        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        val output = ByteArray((cleaned.length * 5) / 8)
        var buffer = 0
        var bitsLeft = 0
        var index = 0

        for (ch in cleaned) {
            val value = alphabet.indexOf(ch)
            if (value < 0) return null
            buffer = (buffer shl 5) or value
            bitsLeft += 5
            if (bitsLeft >= 8) {
                output[index++] = ((buffer shr (bitsLeft - 8)) and 0xFF).toByte()
                bitsLeft -= 8
            }
        }

        return output.copyOfRange(0, index)
    }
}
